include/rapidjson/document.h

```diff
@@ -952,12 +956,16 @@ class GenericValue {
         if (IsUint64()) {
             uint64_t u = GetUint64();
             volatile double d = static_cast<double>(u);
-            return static_cast<uint64_t>(d) == u;
+            return (d >= 0.0)
+                && (d < (static_cast<double>(std::numeric_limits<uint64_t>::max()/2 + 1)) * 2)
```

Comment:
The issue is UINT64_MAX can't be exactly represented in a double. I changed it to be strictly less than UINT64_MAX as a double and added some more tests.